### ISSUE TITLE
Thread/"Respond to single post" option

### ISSUE THREAD

**Issue by theDyingMountain** [nextcloud/spreed#1469]
**Is your feature request related to a problem? Please describe.**
In long chats with many messages and participants, it would be helpful to be able to respond directly to a single post. On slack.com this is called "Start a thread".

**Describe the solution you'd like**
A button "Start thread" which opens a discussion about the post in the right sidebar.
Example from slack:
![grafik](https://user-images.githubusercontent.com/4446224/51686410-9d1c4b00-1ff0-11e9-9112-8b119ac4f171.png)

In the main chat an intented message with the number of replies is shown.
Example from slack:
![grafik](https://user-images.githubusercontent.com/4446224/51686280-66463500-1ff0-11e9-8b44-052541d54d09.png)

**Comment by nickvergessen** [comment]
We will not start threading, but replying is certainly an option.

**Comment by nickvergessen** [comment]
Actually this is duplicate of #1136

**Comment by szaimen** [comment]
@nickvergessen is this really a duplicate of #1136?
I wonder if you implemented threads or just replying to other chat messages?
Because Replying in e.g. in Slack and Whatsapp isn't solved the same way: in Whatsapp you just quote (and link) to the comment you were responding to and in Slack you can create a Thread. Are both ways already implemented with https://github.com/nextcloud/spreed/pull/2000?

**Comment by nickvergessen** [comment]
No we only implemented replying similar to whatsapp. But we are not going to implement threading. We are a chat-focus app, not a forum replacement.

**Comment by szaimen** [comment]
This is not about building a forum replacement. It is about making Talk as
userfriendly and productive as possible.
And threads are an important part of that imo.

Furthermore, just because other chat apps like WhatsApp and Telegram
haven't implemented threads yet doesn't mean that this feature wouldn't be
useful at all and could also differentiate Nextcloud Talk from them in
providing better tools for productivity!

BTW: I think both ways to reply have their right to exist so I think they
should be implemented both.

And one additional level would be enough imo. The possibility to make a new
thread to an answer from a thread isn't necessarily needed. E.g. Microsoft
Teams is doing that just right.

Joas Schilling <notifications@github.com> schrieb am Mo., 11. Nov. 2019,
08:52:

**Comment by szaimen** [comment]
> Replies are not yet released, so I would propose to check it out and rediscuss it again after you tried to use it sometime, although I don't see much point in this.

I will let you know. 👍

**Comment by szaimen** [comment]
BTW:
This is how Threads are handled in Microsoft Teams which is the inspiration for this and works quite well:
![image](https://user-images.githubusercontent.com/42591237/68680960-d5325180-0562-11ea-947a-283a59343a3b.png)
(and would also work quite well in Nextcoud Talk, I am sure!)

**Comment by amsel54** [comment]
I would highly recommend such a feature, to have the possibility to group messages and to collapse and hide such "threads/topics" within a group chat.

**Comment by MPStudyly** [comment]
In theory it should already be possible to implement some sort of threads "light" in the frontend only by leveraging the logic introduced with replies. In this case, threads are not logically different from replies, they are just visualized differently. Telegram implemented such a feature for channels/large groups somewhen during the last two years. When you tap a message (Android) which got replies or is a reply, the context menu has an option "show replies", which opens a new view, showing all related messages just like a thread. When implemented like this, it could be made toggleable by users to keep the linear view (status quo) or visual grouping to threads.

**Comment by small1** [comment]
From what i have seen the last year this is atcually a requested feature. And some opts for apps liker Mattermost or Rocket chat  as talk lacks it.  ar least the basic stuff added here from @szaimen would be enough (rocket.chat has for example threads and discussions that differs in how it handles stuff)   And then we also have slack with their huddle stuff that when you have a smaller video meeting a thread is opened for just that meeting and not cluttering the rest of the chat. This would be good together with the silent meetings just inviting a few participants.

**Comment by fragger1991** [comment]
Meanwhile we also use Mattermost with the thread function. With calls and the focal board, other similar functions were added to Mattermost, which we need in everyday work. Nextcloud was supposed to be our central platform, but due to the lack of threads alone, Nextcloud unfortunately wrote itself off. We currently only use Nextcloud to share files, the clients are used for this, no longer the interface. Too bad that Nextcloud missed this opportunity, Mattermost has now seized it.

**Comment by marcoambrosini** [comment]
more references:
- https://slack.design/articles/threads-in-slack-a-long-design-journey-part-1-of-2/
- https://slack.design/articles/threads-in-slack-a-long-design-journey-part-2-of-2/

**Comment by sascha08-15** [comment]
Is this topic on the roadmap?

**Comment by stef-bloom** [comment]
Strange that nextcloud email forces threads with no option to remove them, while for Talk we have the clutter that comes with unthreaded conversations. 
Please add a feature to reply to messages in public threads, the way I can in Slack.

**Comment by jospoortvliet** [comment]
> Is this topic on the roadmap?

Yes, we'll implement it, one way or another. I don't know the ETA on it, but such a feature is often requested and makes sense. We just have to find a way to do it that works nicely in our user interface. There are some bad implementations out there, thoughts on and examples of good ones is genuinely welcome!

**Comment by githubkoma** [comment]
Dear Team,
thanks for the awesome Talk App !

I love and hate threads equally in other messengers, so i would like to propose an  usability and readability option.

Background (or "what drives me mad"):
Everybody knows the flow of conversations, they have the tendency to get unstructured and important threads of messages get "buried" below new incoming messages in the parent channel.
Also People sometimes dont answer in threads, either they dont want to use threads, or they accidentally answer in the parent channel.
This makes things more than often confusing to follow..

How it works in other messengers now:
   - Open Channel
   - Write either directly into the channel, or reply to a thread

My Proposal to implement perhaps as follows:
 -   Open Channel
 -   DONT ALLOW to write normally. input field is disabled
 -   Just allow to reply to threads, that are already open
 -   AND have the ability to create a new thread
 -  Allow like up to 3-6 Threads per Conversation max, so users who open the channel dont get excessively overwhelmed.
   - Design: Perhaps let it look like Tiles within a Conversation, that can be tapped/clicked on to open up the Thread, and those tiles could have in itself an Unread Counter, ...

Perhaps with an option per Conversation, like in Conversation A this is activated, and in Conversation B,C,D its just like it was before. This way it could be a feature to be experimented with (A/B Testing, starting in Web Version, ...)

Thanks for considering and wishing you the best,
githubkoma

**Comment by githubkoma** [comment]
I made a mockup of what a thread enabled channel perhaps could look like to (1.) achieve or even (2.) enforce a structure:

![threads mockup](https://github.com/nextcloud/spreed/assets/5107805/39b4dfd0-4614-411e-8f57-ea4435600ee2)

**Comment by KaliPhobos** [comment]
Only using the reply-to option doesn't properly scale with user numbers. Especially when multiple concurrent discussions take place in a single group chat, it's very hard to follow -> Threads
I really like Discord's implementation of threads, which allows user-managed "channels" for individual topics, which reduces moderation effeorts (having to create new group chats for new topics). It's especially helpful how users can see a list of "active threads" to pick from to partake in the discussions.
A similar "active threads" overview would be great :)

**Comment by 9joshua** [comment]
Conversation threads the only feature I really need in NextCloud. Following the tested and proven example of Slack threads is the best solution IMO...
![Image](https://github.com/user-attachments/assets/da20cb61-fea2-4b20-bfe0-d126ba914411)

Creating separate conversations for micro topics isn't practical. Take for example, the comment from this [forum post](https://help.nextcloud.com/t/configuration-of-threads-in-talk/201385/2):
> The idea you have one conversation per topic which would loosely translate to a “thread”. You discuss “Ferrari” in one chat and “Lamborgini” in another even same people participate in both chats.

Someone in the Ferrari conversation might mention they have a 1989 Ferrari. Another user does as well and they go on at length about their experiences while flooding the main Ferrari conversation.

**Comment by eldelacajita** [comment]
I think an important part of implementing this type of in-chat threads is _finding_ existing threads, so they don't end up buried up in the chat.

Just a possible solution using the existing Nextcloud interface paradigms: the **right sidebar** (which has tabs like "participants", "shared items" or even "breakout rooms" in conferences), could get a new tab called "threads". To make it relevant, I think it should be the default tab, and always open by default if the chat has threads.

I think that UI choices that focus on avoiding "orphan" mesages that could have joined an existing thread (or duplicated threads), is crucial for the feature to be useful at all.

**Comment by sstcvetkov** [comment]
Hello

Is the threads feature planned to be implemented in the near future? It's 2025 outside...

I consult C-level staff and usually recommend them to choose Nextcloud when the companies really interested in the privacy of users' data. 
Unfortunately, the lack of "team mentioning" and "threads" features is the common reason for using some other chat solutions. This leads to choosing some other high-integrated solutions, but not Nextcoud.
These features - "team mentioning" and "threads" are really important for small+ organizations.

**Comment by rcgch** [comment]
It is a great pity that this thread feature is still missing. It makes it extremely difficult to find leaders who want to give up Slack/Teams/BeeKeeper in favor of Talk.

**Comment by dvgitit** [comment]
This feature would be very helpful or is necessary to keep discussion organized by topic.

**Comment by HolgerHees** [comment]
We are evaluating Nextcloud talk currently and the most missing thing for 90% of all users (~2000) are the missing Thread handling.

As an example, we have some group chats where ~15-20 are invited. Sometimes, more than one topic is discussed in parallel. You can't handle these by creating "Nextcloud topics" inside these chat. You don't know what becomes a thread when you write your first message inside these chat. 

Normally you write a message, another person is answering on this message and more and more people answering. This is how a message should become a thread. All of these happens inside the same group chat.

btw. What exactly is the reason for closing this issues?

**Comment by shu0406** [comment]
> We are evaluating Nextcloud talk currently and the most missing thing for 90% of all users (~2000) are the missing Thread handling.
> 
> As an example, we have some group chats where ~15-20 are invited. Sometimes, more than one topic is discussed in parallel. You can't handle these by creating "Nextcloud topics" inside these chat. You don't know what becomes a thread when you write your first message inside these chat.
> 
> Normally you write a message, another person is answering on this message and more and more people answering. This is how a message should become a thread. All of these happens inside the same group chat.
> 
> btw. What exactly is the reason for closing this issues?

Threads are available since Version Hub10 (32.0.0), I think: 
https://nextcloud.com/blog/november-maintenance-updates-nextcloud-hub-10-nextcloud-hub-25-autumn/

**Comment by nickvergessen** [comment]
- @HolgerHees this is already requested/tracked in https://github.com/nextcloud/spreed/issues/15877

**Comment by HolgerHees** [comment]
We are evaluating 32.0.0 and we already found that you can create Threads.

But again, When you write the first message, you don't know if it will become a thread. It depends on how many people are answering.

Also the solution to convert Messages to Threads, will never work in scenarios like this. People simply expect messages to be grouped or displayed as threads. It's just what they're used before e.g. in Slack , Mattermost or Teams. They don't want to decide at some point if they have to manual convert something to a thread. Maybe some people will do it and some people not. It will never be reliable and will be a chaos after a while.

I'm currently pushing hard to implement Nextcloud in as many areas as possible. But for the majority of users, this is a real deal-breaker. Currently we are using Mattermost and if you click on a Message, you get the full Thread related to this Message.

**Comment by nickvergessen** [comment]
> We are evaluating 32.0.0

That version is outdated already. This week 32.0.4 is coming out :) 

> It's just what they're used before e.g. in Slack, … or Teams.

Cool that you mention it.
- In Slack you can never respond without it being a thread. We will bring this at some point via https://github.com/nextcloud/spreed/issues/13562 and then you can simply take away the permission to post outside of a thread.
- Teams: Is worse. When creating a channel/conversation you have to pick whether it is a "moderated" conversation => Slack style, or whether users can "Reply" normally with quotes being shown, but then you can NOT create a thread at all.

So to summarize: We still plan to split the posting permissions, to allow a "only some people can post messages, everyone else can only reply in threads" case. But apart from that it's basically covered already from that regard.

**Comment by djmaze** [comment]
> In Slack you can never respond without it being a thread. We will bring this at some point via https://github.com/nextcloud/spreed/issues/13562 and then you can simply take away the permission to post outside of a thread.

Will that mean I need to create a message as a thread so anyone can reply to it at all? That seems very.. inconvenient.

**Comment by HolgerHees** [comment]
I would expect that a chat behaves like below.

1. You have a channel with maybe 30 members.
2. Each of them are write several messages into this group channel. In our Company, each team has this kind of group channel. People are just announcing things here, Are asking questions or just saying "Good Morning". Every group channel member can post anything what he thinks is related to this group/team.
4. Sometimes people are answering to a specific message and sometimes, more and more people are answering. This is the moment when it should be automatically provide a thread view, without actively deciding to convert it to a thread. 
 
People don't understand what technicians means with "creating a thread". They just want to write a message. And if someone maybe is answering, they want to see a history of answers related to the current message where they clicked and respond. And they still want to see the global chronological message order in the whole team channel.

Most people read quickly the chat messages of the day, and as soon as a message becomes interesting, they want to see all the answers to it. 

e.g. in Mattermost it behaves exactly like this. You have a chronological order of messages and in the moment when you click on a message, you see all related answers/messages on a dedicated area on the right side.

**Comment by hansimausisteffiwurstbrot** [comment]
Just here to underline the importance of an easy way to enable threading, as in Mattermost.

It's great to have all the different Nextcloud apps that are continuously interlinked. However, not having this feature in group chats is a hindrance for me, too. In several groups, we continue to use Nextcloud as a file sharing platform and do the communication in Mattermost due to this lacking feature.

**Comment by jameskimmel** [comment]
> In Slack you can never respond without it being a thread. 

You say it like this is something negative. IMHO, this is one of the reasons why Slack became so big. 

Teams became big because people already have it preinstalled and already have it in their subscription.
But Slack became big because, despite not being self-hosted, expensive pricing, and not being open, it is 
the best tool for the job. IMHO, there is no shame in learning from the best.
